x-pack/legacy/plugins/apm/server/lib/service_map/run_service_map_task.ts

```diff
@@ -0,0 +1,250 @@
+              shard_size: 20,
+              script: {
+                lang: 'painless',
+                source: "(int)doc['span.duration.us'].value/100000"
+              }
+            },
+            aggs: {
+              tracesample: {
+                top_hits: {
+                  size: 20,
+                  _source: ['trace.id', '@timestamp']
+                }
+              }
+            }
+          }
+        }
+      }
+    }
+  };
+
+  if (afterKey) {
+    query.aggs['ext-conns'].composite.after = afterKey;
+  }
+  return query;
+}
+
+function findConns(traceIds: string[]) {
+  const query = {
+    size: 0,
+    query: {
+      bool: {
+        should: [
+          { exists: { field: 'span.id' } },
+          { exists: { field: 'transaction.type' } }
+        ] as any[],
+        minimum_should_match: 2
+      }
+    },
+    aggs: {
+      trace_id: {
+        terms: {
+          field: 'trace.id',
+          order: { _key: 'asc' },
+          size: traceIds.length
+        },
+        aggs: {
+          connections: {
+            scripted_metric: {
+              init_script: 'state.spans = new HashMap();',
+              map_script: { id: 'map-service-conns' },
+              combine_script: { id: 'combine-service-conns' },
+              reduce_script: { id: 'reduce-service-conns' }
+            }
+          }
+        }
+      }
+    }
+  };
+
+  for (const tid of traceIds) {
+    query.query.bool.should.push({ term: { 'trace.id': tid } });
+  }
+  return query;
+}
+
+interface Service {
+  name: string;
+  environment?: string;
+}
+
+interface ConnectionDoc {
+  '@timestamp': string;
+  observer: { version_major: number }; // TODO: make this dynamic
+  service: Service;
+  callee?: Service;
+  connection: Connection;
+  destination?: { address: string };
+}
+
+interface Connection {
+  upstream: { list: string[]; keyword: string };
+  in_trace: string[];
+  type: string;
+  subtype?: string;
+}
+
+export async function runServiceMapTask(
+  server: Server,
+  config: any,
+  lastRun?: string
+) {
+  const apmIdxPattern = config.get('apm_oss.indexPattern');
+  const serviceConnsDestinationIndex = config.get(
+    'xpack.apm.serviceMapDestinationIndex'
+  );
+  const serviceConnsDestinationPipeline = config.get(
+    'xpack.apm.serviceMapDestinationPipeline'
+  );
+
+  const callCluster = server.plugins.elasticsearch.getCluster('data')
+    .callWithInternalUser;
+
+  let mostRecent = '';
+  let afterKey = null;
+  while (true) {
+    const q = interestingTransactions(lastRun, afterKey);
+    const txs = await callCluster('search', {
+      index: apmIdxPattern,
+      body: q
+    });
+
+    if (txs.aggregations['ext-conns'].buckets.length < 1) {
+      return { mostRecent };
+    }
+    afterKey = txs.aggregations['ext-conns'].after_key;
+
+    const traces = new Set<string>();
+
+    txs.aggregations['ext-conns'].buckets.forEach((bucket: any) => {
+      bucket.smpl.tracesample.hits.hits.forEach((hit: any) => {
```

Comment:
We should have proper types here

Reply:
I'm adding support for `composite` aggregations in https://github.com/elastic/kibana/pull/51612. Should be possible for @ogupte to copy and paste the relevant bits, LMK if that doesn't work out.

Reply:
i already added support for `composite`, `diversified_sampler`, and `scripted_metric` aggregations in this PR

Reply:
I'll definitely replace my composite stuff with yours, your key inference it very nice :)